File: code/프로그래머스/모음사전.java
public class 모음사전 {
    StringBuilder words = new StringBuilder();
    int answer = 0;
    boolean find = false;
    String[] array = new String[]{"A", "E", "I", "O", "U"};
    
    public void back_tracking(String word) {
        int length = words.length();
        
        if (length >= 5)
            return;
            
        for(String w : array) {
            words.append(w);
            answer += 1;
            if (word.equals(words.toString())) {
                find = true;
                return;
            }
            back_tracking(word);
            if (find) {
                return;
            }
            words.deleteCharAt(length - 1);
            
        }
    }
    
    public int solution(String word) {
        back_tracking(word);
        return answer;
    }
}

File: code/프로그래머스/모음사전.py
# https://school.programmers.co.kr/learn/courses/30/lessons/84512

import sys

word = []
words = []
array = ['A', 'E', 'I', 'O', 'U']

def back_tracking():
    global word, words, array
    length = len(word)
    if length >= 5:
        return
    for w in array:
        word.append(w)
        words.append(''.join(word))
        back_tracking()
        word.pop()
        

def solution(word):
    sys.setrecursionlimit(10000000)
    back_tracking()
    
    answer = words.index(word)
    return answer + 1

solution('AAAAE')